File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10**6)

# 가장 긴 증가하는 부분 수열 = 최장 부분 증가수열
n = int(input())
data = list(map(int, input().split()))
data.insert(0,-1)
dp = [0] * (n+1)
dp[1] = 1 #자명한 사실

for i in range(2,n+1):
    max_length = 0
    for j in range(1,i): #본인 앞에까지 중에
        if dp[j] > max_length and data[j] < data[i]: #앞에 숫자가 작아야 현재 숫자가 그 뒤에 붙을 수 있지
            max_length = dp[j]
    dp[i] = max_length + 1

print(max(dp))
            

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {

    static int n;
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int[] data, dp;
    static int res;
    public static void main(String[] args) throws IOException {
        n = Integer.parseInt(br.readLine());
        data = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++) {
            data[i] = Integer.parseInt(st.nextToken());
        }
        dp = new int[n];
        dp[0] = 1;  // 자명한 사실
        res = 1;
        for (int i = 1; i < n; i++) {
            int max_length = 0;
            for (int j = 0; j < i; j++) {
                if (data[j] < data[i] && max_length < dp[j]) {
                    max_length = dp[j];
                }
            }
            dp[i] = max_length + 1;
            res = Math.max(res, dp[i]);
        }
        System.out.println(res);
    }
}